backend/src/apiserver/resource/resource_manager.go

```diff
@@ -195,616 +510,1147 @@ func (r *ResourceManager) ArchiveExperiment(ctx context.Context, experimentId st
+		osErr = util.Wrapf(err, "Failed to delete pipeline spec for pipeline version id %v with URI %v", pipelineVersionId, pipelineVersion.PipelineSpecURI)
+	} else {
+		pipelineSpecRemoved = true
 	}
-	namespace, err := r.GetNamespaceFromRunID(runID)
+	// Delete based on pipeline version id
+	err = r.objectStore.DeleteFile(r.objectStore.GetPipelineKey(fmt.Sprint(pipelineVersionId)))
 	if err != nil {
-		return util.Wrap(err, "Delete run failed")
+		glog.Errorf("%v", util.Wrapf(err, "Failed to delete pipeline spec for pipeline version id %v", pipelineVersionId))
+		err = util.Wrapf(err, "Failed to delete pipeline spec for pipeline version id %v", pipelineVersionId)
+		osErr = util.Wrap(osErr, err.Error())
+	} else {
+		pipelineSpecRemoved = true
 	}
-	err = r.getWorkflowClient(namespace).Delete(ctx, runDetail.Name, v1.DeleteOptions{})
+	// Delete based on pipeline id
+	err = r.objectStore.DeleteFile(r.objectStore.GetPipelineKey(fmt.Sprint(pipelineVersion.PipelineId)))
 	if err != nil {
-		// API won't need to delete the workflow CR
-		// once persistent agent sync the state to DB and set TTL for it.
-		glog.Warningf("Failed to delete run %v. Error: %v", runDetail.Name, err.Error())
+		glog.Errorf("%v", util.Wrapf(err, "Failed to delete pipeline spec for pipeline version id %v using pipeline id %v", pipelineVersionId, pipelineVersion.PipelineId))
+		err = util.Wrapf(err, "Failed to delete pipeline spec for pipeline version id %v using pipeline id %v", pipelineVersionId, pipelineVersion.PipelineId)
+		osErr = util.Wrap(osErr, err.Error())
+	} else {
+		pipelineSpecRemoved = true
 	}
-	err = r.runStore.DeleteRun(runID)
+	if !pipelineSpecRemoved {
+		return util.Wrap(osErr, "Failed to delete a pipeline spec")
+	}
+	// Delete the DB entry
+	err = r.pipelineStore.DeletePipelineVersion(pipelineVersionId)
 	if err != nil {
-		return util.Wrap(err, "Delete run failed")
+		glog.Errorf("%v", util.Wrapf(err, "Failed to delete a DB entry for pipeline version id %v", pipelineVersionId))
+		return util.Wrapf(err, "Failed to delete a DB entry for pipeline version id %v", pipelineVersionId)
 	}
 	return nil
 }
 
-func (r *ResourceManager) CreateTask(ctx context.Context, apiTask *apiv1beta1.Task) (*model.Task, error) {
-	uuid, err := r.uuid.NewRandom()
+// Creates a run and schedule a workflow CR.
+// Note: when creating a run from a manifest, this triggers creation of
+// a new pipeline and pipeline version that share the name, description, and namespace.
+// If run.ExperimentId is not specified, it is set to the default experiment.
+// If run.Namespace  is no specified, it gets inferred from the parent experiment.
+// Manifest's namespace gets overwritten with the run.Namespace.
+// Creating a run from recurring run prioritizes recurring run's pipeline spec over the run's one.
+func (r *ResourceManager) CreateRun(ctx context.Context, run *model.Run) (*model.Run, error) {
```

Comment:
This function became really long after the change (from ~70 lines to almost 200 lines), which really affects its readability. Maybe try splitting it into smaller subfunctions?

Reply:
That's a good point. Let me try improving it.